Re-parse Command

jtimpe opened the pull request:
## Summary of Changes
- Introduced new command that allows an admin user to re-parse a subset or all of the datafiles in the database. 
  -  Allows admin to specify whether to re-parse in place or to create new indices
  - Creates DB backup
- Created Django command that parametrizes the DB backup routine
Pull request closes #3004 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Submit some datafiles for different program types,  fiscal years, and fiscal quarters.
3. Open http://localhost:9200/_cat/indices/?pretty&v&s=index and verify the elastic indices match the DAC
4. Execute the `clean_and_reparse` Django command a myriad of times with different values for the parameters.
  - Between each run of the command verify with the Elastic url above and the DAC that you see what you would expect.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Postgres record counts never change/Postgres always remains source of truth
- [x] Elastic indices match with parameters passed to re-parse command
- [x] App can archive/delete live data set
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary
- [x] `lfrohlich` and/or `adpennington` confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

jtimpe committed:
add clean and reparse mgmt cmd

jtimpe committed:
wip new clean and reparse

jtimpe committed:
spruce up clean and reparse

jtimpe committed:
revert to original index naming scheme

jtimpe committed:
local vars attempt

elipe17 committed:
- Added logic for local DB backup and restore

elipe17 committed:
- Added command that supports local and remote DB backup/restore
- Updated reparse command to leverage backup/restore command

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3004-clean-and-reparse-cmd

elipe17 committed:
- adding logging throughout

elipe17 committed:
- Added new args for command
- Added final backup check just in case
- Added password to cmd shell context
- Updated index naming convention and logging

elipe17 committed:
- Updated command to delete documents if we aren't creating a new set of indices

elipe17 committed:
- aligning backup timestamp with elastic index timestamp

elipe17 committed:
- Remove unnecessary restore option from command. Can't be done from Django command.
- Added continue message to repasrse command
- lint

elipe17 committed:
- updated `handle` method to not be so linter complex

elipe17 committed:
- Rename command to align with functionality

elipe17 committed:
- add missing docstrings

elipe17 committed:
- move file size check to local backup piece of code

andrew-jameson commented:
Step 4 in the 'How to Test' section needs a little more verbosity and usage examples for QASP unless a markdown document is planned.

andrew-jameson commented on tdrs-backend/tdpservice/scheduling/db_backup.py:
```diff
@@ -142,7 +142,7 @@ def restore_database(file_name, postgres_client, database_uri, system_user):
     except Exception as e:
         logger.error(f"Caught exception while creating the database. Exception: {e}.")
-        return False
+        raise e
```
Why bother with the try/except if we're going to raise anyway? The logging line doesn't currently add value to the troubleshooting process.

elipe17 replied:
Mostly for consistency with the rest of the functions in this file. Every other function in the file follows that needs exception handling follows that pattern. Also, most of those functions are called from different processes so it helps a little with traceability imo.

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+
+        c = str(input(f'\n{continue_msg}\nContinue [y/n]? ')).lower()
+        if c not in ['y', 'yes']:
+            print('Cancelled.')
```
Logger needed or line no longer needed?

elipe17 replied:
This needs to be generic std::out for the user perspective. Doesn't make a lot of sense to have a logger write this from the user perspective.

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+            continue_msg = continue_msg.format(fy="All", q="Q1-4")
+        else:
+            if not fiscal_year and not fiscal_quarter:
+                print(
```
`print` intended vs logger?

elipe17 replied:
Same as above. It needs to be std::out to indicate to the user that something is wrong, not to the developer.

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+                backup_file_name += f"_FY_All_Q{fiscal_quarter}"
+                continue_msg = continue_msg.format(fy="All", q=fiscal_quarter)
+
+        continue_msg += "will {new_index} stored in new indices and the old indices ".format(new_index="be"
+                                                                                             if new_indices
+                                                                                             else "NOT be")
+        continue_msg += "will {old_index} deleted.".format(old_index="be" if delete_indices else "NOT be")
+        continue_msg += "\nThese options will delete and re-parse {0} datafiles.".format(f"ALL ({files.count()})" if
+                                                                                         delete_all else
+                                                                                         f"({files.count()})")
```
Indentation and if expressions should be altered for readability.

andrew-jameson commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+                files = files.filter(year=fiscal_year, quarter=fiscal_quarter)
+                backup_file_name += f"_FY_{fiscal_year}_Q{fiscal_quarter}"
+                continue_msg = continue_msg.format(fy=fiscal_year, q=fiscal_quarter)
+            elif fiscal_year:
```
Conditionals should follow Pythonic explicitness (e.g., `elif fiscal_year is True` or `elif fiscal_year exists) as the initialization above can result in non-Boolean types (None, String, Integer).

elipe17 committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

raftmsohani commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+            log("Database backup FAILED. Clean and re-parse NOT executed. Database and Elastic are CONSISTENT!",
+                logger_context=log_context,
+                level='error')
+            raise e
```
You can just log the error and then return

elipe17 replied:
I personally think it is very important that this command raises an exception in any situation that it runs into a failure to be extremely explicit about the fact that it failed and the database and elastic could be very inconsistent. I know that isn't the case with this exception, but it sets the precedence for consistency across the command. Maybe @andrew-jameson  or @jtimpe could provide some extra feedback?

andrew-jameson replied:
I think I agree with @elipe17 that we should raise whatever the exception is here.

raftmsohani commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+            # atomic delete?
+            try:
+                model = doc.Django.model
+                qset = model.objects.all().filter(datafile_id__in=file_ids)
```
I guess you can move the query to outside the loop?!!

elipe17 replied:
Could you explain how I might be able to move the query out of the loop?

raftmsohani commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+                    doc().update(qset, refresh=True, action='delete')
+                qset._raw_delete(qset.db)
+            except BulkIndexError as e:
+                log(f'Elastic document delete failed for model {model}. Database and Elastic are INCONSISTENT! Restore '
```
nit: you should be getting same result with log.error

raftmsohani commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+    def __handle_datafiles(self, files, log_context):
+        """Delete, re-save, and re-parse selected datafiles."""
+        for f in files:
+            try:
```
you can put both delete and save in one try/catch and catch two separate exceptions

raftmsohani commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,236 @@
+                    'Provide either option to continue, or --all to wipe all submissions.'
+                )
+                return
+            files = DataFile.objects.all()
```
not sure why you will need to do this again, if this is working, then you should already have this from try

elipe17 committed:
- Added better exception handling
- Updated clearing logic on sorted cases to maintain state better for deletes
- updated test

elipe17 committed:
- updated based on feedback
- moved all code into exception handling on the restore function
- Added acceptable choices for quarter
- Genericized exception handling
- updated string formatting

elipe17 committed:
- linting

elipe17 committed:
Merge branch '3004-clean-and-reparse-cmd' of https://github.com/raft-tech/TANF-app into 3004-clean-and-reparse-cmd

elipe17 committed:
- intermediate commit for md

elipe17 committed:
- updated md

elipe17 committed:
- add usage when no args

elipe17 committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

elipe17 committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

elipe17 committed:
- fixed logic error

elipe17 committed:
Merge branch '3004-clean-and-reparse-cmd' of https://github.com/raft-tech/TANF-app into 3004-clean-and-reparse-cmd

elipe17 committed:
- Adding another celery worker and making each worker process two tasks concurrently
- using correct db name

elipe17 committed:
- Use correct backup logic

elipe17 committed:
remove backup

elipe17 committed:
- fix lint

elipe17 committed:
- Updating parser errors to be raw deleted because we need the speed
- added todo in backup script
- added extra logging in datafile function
- moved task.delay into exception block

elipe17 committed:
- Reverting back to one worker one thread in celery
- Move error deletion to a better spot
- Force elastic log level to WARN always

elipe17 committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

andrew-jameson commented on tdrs-backend/gunicorn_start.sh:
```diff
@@ -12,16 +12,17 @@ else
 
 
-celery -A tdpservice.settings worker -c 1 &
+celery -A tdpservice.settings worker --loglevel=WARNING --concurrency=1 -n worker1@%h &
```
What do the special symbols ending of `worker1@%h` indicate?

elipe17 replied:
It is a syntax that celery recommends to associate each worker with the host it is running on. In the future if/when we add more workers it will be a little easier to delineate workers having issues. Especially if we allow workers to span multiple hosts.

raftmsohani replied:
could you add the link as a docstring for future ref?

andrew-jameson commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -236,15 +236,20 @@ def add_record(self, case_hash, record_doc_pair, line_num):
             self.hash_sorted_cases = dict()
-            self.cases = dict()
+
+            # We don't want to re-assign self.cases here because we lose the keys of the record/doc types we've already
+            # made. If we don't maintain that state we might not delete everything if we need to roll the records back
+            # at the end of, or during parsing.
+            for key in self.cases.keys():
+                self.cases[key] = {}
```
Great comment explaining why we need to delete just the contents but not the keys/structure.

andrew-jameson commented on tdrs-backend/tdpservice/scheduling/parser_task.py:
```diff
@@ -37,10 +37,11 @@ def parse(data_file_id):
-        account_approval_status=AccountApprovalStatusChoices.APPROVED,
-        groups=Group.objects.get(name='Data Analyst')
-    ).values_list('username', flat=True).distinct()
+    if should_send_submission_email:
```
Minor nit, but to be pythonic we should have an explicit comparison such as `if should_send_submission_email is True:`. In the past we've been bit using an implicit conditional evaluating when an edge case or error presented itself.

andrew-jameson commented on tdrs-backend/tdpservice/scheduling/db_backup.py:
```diff
@@ -140,32 +149,32 @@ def restore_database(file_name, postgres_client, database_uri, system_user):
+        logger.info(msg)
+        return True
+    except Exception as e:
+        logger.error(f"Caught exception while creating the database. Exception: {e}.")
```
```suggestion
        logger.error(f"Caught exception while restoring the database. Exception: {e}.")
```

I know it's not really your scope, but after reviewing this expanded try/except, felt this was more apt description.

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -119,7 +117,7 @@ def bulk_create_records(unsaved_records, line_number, header_count, datafile, df
 
                 try:
```
nit: we should try merging this try with the upper try and catch specific exception. No need to have two tries.

elipe17 committed:
- Made If statement pythonic
- Updated log message
- condensed try except blocks to one block

elipe17 committed:
- fix lint

elipe17 committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

raftmsohani commented on tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:
```diff
@@ -0,0 +1,249 @@
+                    logger_context=log_context,
+                    level='critical')
+                raise e
+            except Exception as e:
```
For readability, my suggestion is to write an exception for raw query instead of using generic Exception. Using Exception should be the last resort

elipe17 replied:
I do disagree with this in this case. I think we need to catch the lowest possible exception case to ensure we notify the user that a failure occurred and the DB could be in an inconsistent state. Im open to adding more exceptions to the `except` block, but I think every single `except` block should be catching `Exception` as a safety net.

raftmsohani replied:
What I mean is catch as much specific exceptions before catching generic "Exception"

elipe17 replied:
I added some slightly more specific exception handling. Let me know what you think!

raftmsohani commented on tdrs-backend/tdpservice/scheduling/db_backup.py:
```diff
@@ -94,7 +94,16 @@ def backup_database(file_name,
+        db_name = settings.DATABASES['default']['NAME']
+        db_user = settings.DATABASES['default']['USER']
+
+        export_password = f"export PGPASSWORD={settings.DATABASES['default']['PASSWORD']}"
```
I have been doing the re-parsing and it always fails at the db_backup step. Happy to pair on this

elipe17 committed:
- Updated dockerfile to use pg client 15

elipe17 committed:
- Adding link to docs

elipe17 committed:
- Added slightly more specific exception handling

jtimpe commented on tdrs-backend/tdpservice/search_indexes/management/commands/backup_db.py:
```diff
@@ -0,0 +1,53 @@
+import os
+from django.core.management.base import BaseCommand
+from django.conf import settings
+from tdpservice.scheduling.db_backup import main, get_system_values
```
might be best to co-locate this management command w/ the rest of the backup script

elipe17 replied:
I like that. I made the changes!

jtimpe reviewed (commented):
@elipe17 it doesn't let me approve as a pr creator, but all the changes you made to get the database backup working as well as the refined input/exception handling are excellent. APPROVED, lgtm 🚀

elipe17 committed:
- re located backup command with backup code.

andrew-jameson commented on docs/Technical-Documentation/clean-and-reparse.md:
```diff
@@ -0,0 +1,275 @@
+    - The primary key for all reparsed datafiles should no longer be the same.
+    - `ParserError` and `DataFileSummary` objects should be consistent with the file.
+
+- First run:
```
Let's cut the long outputs from these runs. The initial local output starting L40 will give the reader a sense of what to look at/for.

andrew-jameson commented on docs/Technical-Documentation/clean-and-reparse.md:
````diff
@@ -0,0 +1,275 @@
+2. The expected results for this command will be exactly the same as above. The only difference is that no matter how many times you execute this command, you should only see 21 indices in Elastic with the `dev` prefix.
+
+```
+health status index                                              uuid                   pri rep docs.count docs.deleted store.size pri.store.size
````
Remove output section here.

andrew-jameson commented on docs/Technical-Documentation/clean-and-reparse.md:
````diff
@@ -0,0 +1,275 @@
+- While the DAC reports the correct number of records for all submitted types, Elastic does not. This is because we only reparsed a subset of the entire collection of datafiles for the first time we executed the `clean_and_reparse` command. Therefore, Elastic only has documents for the subset of resubmitted files. If we had already executed the command: `python manage.py clean_and_reparse -a -n`  ...
+
+```
+health status index                                              uuid                   pri rep docs.count docs.deleted store.size pri.store.size
````
For clarity/conciseness, perhaps we can remove the columns that aren't docs.count and any records with a count of zero for these comparison sections?

andrew-jameson reviewed (changes requested):
requested some minor feedback on the markdown, will approve once eric re-requests review or dm's me

elipe17 committed:
- Updated elastic output to be more brief

raftmsohani reviewed (approved):
LGTM! Tried a couple of edge cases and it worked. Great Work! 🥇

elipe17 committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

ADPennington committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

ADPennington committed:
Merge branch 'develop' into 3004-clean-and-reparse-cmd

ADPennington commented on tdrs-frontend/.env:
```diff
@@ -37,7 +37,8 @@ REACT_APP_DEBOUNCE_TIME=30000
 # Enable the Kibana tab for dev purposes.
-# REACT_APP_DEV_KIBANA=true
+# TODO: Comment once done developing
+REACT_APP_DEV_KIBANA=true
```
Reminder to comment out once qasp-approved.

ADPennington commented on docs/Technical-Documentation/clean-and-reparse.md:
```diff
@@ -0,0 +1,129 @@
+As a safety measure, this process must ALWAYS be executed manually by a system administrator. Once executed, all processes thereafter are completely
+automated. The steps below outline how this process executes.
+
+1. System admin logs in to the appropriate backend application. E.g. `tdp-backend-raft`.
```
because this feature currently requires admins to work from CLI, we'll need more documentation on the commands to run on GFE. below are the commands I had to run for step 1. 

```shell=

# 0. disconnect from VPN. 

# 1. authenticate into cloud.gov 
apennington@HHSLBDSWL73 MINGW64 /
$ cf login -a api.fr.cloud.gov  --sso
API endpoint: api.fr.cloud.gov

Temporary Authentication Code ( Get one at https://login.fr.cloud.gov/passcode ): <one-time passcode redacted>

Authenticating...
OK


Select an org:
1. hhs-acf-ofa
2. sandbox-hhs

Org (enter to skip): 1
1
Targeted org hhs-acf-ofa.

Select a space:
1. tanf-dev
2. tanf-prod
3. tanf-staging

Space (enter to skip): 1
1
Targeted space tanf-dev.

API endpoint:   https://api.fr.cloud.gov
API version:    3.170.0
user:           alexandra.pennington@acf.hhs.gov
org:            hhs-acf-ofa
space:          tanf-dev

# 2. ssh into backend app
apennington@HHSLBDSWL73 MINGW64 /
$ cf curl v3/apps/$(cf app tdp-backend-qasp --guid)/processes | jq --raw-output '.resources | .[]? | select(.type == "web").guid'
<PROCESS_GUID redacted>

apennington@HHSLBDSWL73 MINGW64 /
$ cf ssh-code
<ssh code redacted>

apennington@HHSLBDSWL73 MINGW64 /
$ ssh -p 2222 cf:<PROCESS_GUID redacted>/0@ssh.fr.cloud.gov
The authenticity of host '[ssh.fr.cloud.gov]:2222 ([2620:108:d00f::fcd:e8d8]:2222)' can't be established.
RSA key fingerprint is <key redacted>.
This key is not known by any other names
Please type 'yes', 'no' or the fingerprint: yes
Could not create directory '/u/.ssh' (No such file or directory).
Failed to add the host to the list of known hosts (/u/.ssh/known_hosts).
cf:<PROCESS_GUID redacted>/0@ssh.fr.cloud.gov's password:<ssh code will be invisible>
vcap@f8eeb863-ee2b-4b0d-7e7c-bc63:~$

# 3. change to directory where python commands can be run
/tmp/lifecycle/shell

# 4. check re-parse command options
python manage.py clean_and_reparse -h
```

ADPennington commented on docs/Technical-Documentation/clean-and-reparse.md:
```diff
@@ -0,0 +1,129 @@
+7. `clean_and_reparse` deletes documents from appropriate Elastic indices pending the system admin's command choices.
+8. `clean_and_reparse` deletes all Postgres rows associated to all selected datafiles.
+9. `clean_and_reparse` deletes `DataFileSummary` and `ParserError` objects associated with the selected datafiles.
+10. `clean_and_reparse` re-saves the selected datafiles to the database.
```
is there a specific  parameter that is responsible for deleting and replacing the datafile object?

ADPennington replied:
some additional parameters we'd like to have (note: this is follow-on work):

ability to specify:

- file id
- STT
- STT type
- region (i.e. a set of stts in a particular region)
- program type (i.e. tanf or ssp)
- most recent file version 
- processing status (eg pending)
- section

ADPennington replied:
we noticed that most of the original submission history details are retained, except the submission date is overwritten when the reparse command runs (based on the parameters we specify). Is it possible to retain the original submission date? this would be helpful for STT auditors, who use submission history details to confirm if data was submitted on-time.

elipe17 committed:
- Minor updates to log messages
- Fixed minor logic error

elipe17 committed:
- Add ofa login info

elipe17 committed:
- Add missing 'Q'

elipe17 committed:
- recomment var

ADPennington reviewed (approved):
this looks great @elipe17 🚀  thank you!

Notes for awareness and ticket priority-setting @robgendron @lfrohlich @ttran-hub:
- this feature currently operates from CLI. This PR includes documentation for how to run these commands from GFE. Sys admins will pair on this task, for now, but #3106 will track feasibility work for being able to trigger re-parsing from the TDP interfaces.     

- current options facilitate re-parsing primarily by fiscal period, but #3108 will track more options to further customize which files to re-parse (e.g. all tribes, all active files for Washington state, latest version of a file, etc. )

- when the re-parsing command is run, the original submission details (e.g. data submitter, data file) is retained, but the submission date is overwritten. #3107 will address this. 

- the dev team recommends that we hold off on executing this command until #3064 and #3065 are implemented, which will help ensure that sys admins has more visibility into the status of the re-parsing process to help ensure that the db is in a stable state.
